File: java/dsa/data-structures/stack-queue/simplify-path.java
class Solution {
    public String simplifyPath(String path) {
        Stack<String> stack = new Stack<>(); // @viz: init
        String[] parts = path.split("/"); // @viz: split
        
        for (String part : parts) { // @viz: loop
            if (part.equals(".") || part.isEmpty()) { // @viz: check_skip
                continue; // @viz: skip
            }
            
            if (part.equals("..")) { // @viz: check_parent
                if (!stack.isEmpty()) { // @viz: check_pop
                    stack.pop(); // @viz: pop
                }
            } else {
                stack.push(part); // @viz: push
            }
        }
        
        return "/" + String.join("/", stack); // @viz: finish
    }
}

File: javascript/dsa/data-structures/stack-queue/simplify-path.js
function simplifyPath(path) {
  const stack = []; // @viz: init
  const components = path.split("/"); // @viz: split

  for (const part of components) { // @viz: loop
    if (part === "." || part === "") { // @viz: check_skip
      continue; // @viz: skip
    }
    
    if (part === "..") { // @viz: check_parent
      if (stack.length > 0) { // @viz: check_pop
        stack.pop(); // @viz: pop
      }
    } else {
      stack.push(part); // @viz: push
    }
  }

  const finalPath = "/" + stack.join("/"); // @viz: finish
  return finalPath;
}